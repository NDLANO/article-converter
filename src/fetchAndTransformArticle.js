import cheerio from 'cheerio';
import { fetchArticle } from './api/articleApi';
import { transform } from './transformers';
import config from './config';
import { getCopyString } from './plugins/pluginHelpers';

export async function transformArticle(
  article,
  lang,
  accessToken,
  options = {}
) {
  const articleContent = article.content.content
    ? cheerio.load(article.content.content, {
        recognizeSelfClosing: true,
      })
    : undefined;
  const { html, embedMetaData } = articleContent
    ? await transform(
        articleContent,
        lang,
        accessToken,
        article.visualElement,
        options
      )
    : {};
  const copyText = getCopyString(
    article.title.title,
    config.ndlaFrontendDomain,
    options.path,
    article.copyright,
    lang
  );

  return {
    ...article,
    content: html || '',
    metaData: { ...embedMetaData, copyText } || '',
    title: article.title.title || '',
    tags: article.tags ? article.tags.tags : [],
    introduction: article.introduction
      ? article.introduction.introduction
      : undefined,
    metaDescription: article.metaDescription
      ? article.metaDescription.metaDescription
      : '',
  };
}

export default async function fetchAndTransformArticle(
  articleId,
  lang,
  accessToken,
  options = {}
) {
  const article = await fetchArticle(articleId, accessToken, lang);
  const transformedArticle = await transformArticle(
    article,
    lang,
    accessToken,
    options
  );
  return transformedArticle;
}
