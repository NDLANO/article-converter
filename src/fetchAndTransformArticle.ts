import cheerio from 'cheerio';
import { ArticleApiType, fetchArticle } from './api/articleApi';
import { transform } from './transformers';
import config from './config';
import { getCopyString } from './plugins/pluginHelpers';
import { LocaleType, TransformedArticle, TransformOptions } from './interfaces';

export async function transformArticle(
  article: ArticleApiType,
  lang: LocaleType,
  accessToken: string,
  options: TransformOptions = {},
): Promise<TransformedArticle> {
  const articleContent = article.content.content
    ? cheerio.load(article.content.content, {
        recognizeSelfClosing: true,
      })
    : undefined;

  const { html, embedMetaData } = articleContent
    ? await transform(articleContent, lang, accessToken, article.visualElement, options)
    : { html: undefined, embedMetaData: undefined };

  const htmlString: string = html ?? '';

  const copyText: string = getCopyString(
    article.title.title,
    config.ndlaFrontendDomain,
    options.path,
    article.copyright,
    lang,
  );
  const hasContent = article.articleType === 'standard' || cheerio.load(htmlString).text() !== '';

  return {
    ...article,
    content: hasContent ? htmlString : '',
    metaData: { ...embedMetaData, copyText } || '',
    title: article.title.title || '',
    tags: article.tags ? article.tags.tags : [],
    introduction: article.introduction ? article.introduction.introduction : undefined,
    metaDescription: article.metaDescription ? article.metaDescription.metaDescription : '',
  };
}

export default async function fetchAndTransformArticle(
  articleId: string,
  lang: LocaleType,
  accessToken: string,
  options = {},
): Promise<TransformedArticle> {
  const article = await fetchArticle(articleId, accessToken, lang);
  return await transformArticle(article, lang, accessToken, options);
}
