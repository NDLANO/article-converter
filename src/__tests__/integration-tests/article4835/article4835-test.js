/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import { loglessTest, prettify } from '../../testHelpers';
import article4835 from './article4835';

import fetchAndTransformArticle from '../../../fetchAndTransformArticle';

loglessTest('app/fetchAndTransformArticle 4835', async () => {
  nock('http://ndla-api')
    .get('/article-api/v2/articles/4835?language=nb&fallback=true')
    .reply(200, article4835);
  nock('http://ndla-api').head('/files/103066/1p_arsplan_2013-2014_nynorsk.doc').reply(200);
  nock('http://ndla-api').head('/files/103066/1p_arsplan_2013-2014_nynorsk.odt').reply(200);
  nock('http://ndla-api').head('/files/103066/1p_arsplan_2013-2014_nynorsk.pdf').reply(200);

  const transformed = await fetchAndTransformArticle('4835', {
    lang: 'nb',
    accessToken: 'some_token',
    feideToken: 'some_other_token',
  });
  const { content, ...rest } = transformed;

  expect(rest).toMatchSnapshot();
  expect(prettify(content)).toMatchSnapshot();
});
