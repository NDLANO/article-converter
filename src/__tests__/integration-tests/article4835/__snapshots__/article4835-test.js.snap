// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/fetchAndTransformArticle 4835 1`] = `
Object {
  "articleType": "standard",
  "copyright": Object {
    "creators": Array [
      Object {
        "name": "Grete Larsen",
        "type": "Writer",
      },
    ],
    "license": Object {
      "description": "Creative Commons Attribution-NonCommercial-ShareAlike 2.0 Generic",
      "license": "CC-BY-NC-SA-4.0",
      "url": "https://creativecommons.org/licenses/by-nc-sa/2.0/",
    },
    "origin": "",
    "processors": Array [],
    "rightsholders": Array [],
  },
  "created": "2018-06-07T08:44:29Z",
  "id": 4835,
  "introduction": "Forslag til årsplan for faget Matematikk Vg1P. Ligg vedlagt i ulike filformat slik at du kan endre og redigere.",
  "metaData": Object {
    "copyText": "Forfatter: Grete Larsen. Forslag til årsplan [Internett]. Hentet fra: https://test.ndla.no Lest: 19.05.2021",
  },
  "metaDescription": "Forslag til årsplan for faget Matematikk Vg1P. Ligg vedlagt i ulike filformat slik at du kan endre og redigere.",
  "metaImage": Object {
    "alt": "Eit bilde",
    "language": "nn",
    "url": "https://test.api.ndla.no/image-api/raw/id/2061",
  },
  "oldNdlaUrl": "//red.ndla.no/node/27264",
  "requiredLibraries": Array [],
  "revision": 1,
  "supportedLanguages": Array [
    "nb",
    "nn",
  ],
  "tags": Array [
    "framdriftsplan",
    "plan",
    "årsplan",
  ],
  "title": "Forslag til årsplan",
  "updated": "2018-06-07T08:44:31Z",
  "updatedBy": "r0gHb9Xg3li4yyXv0QSGQczV3bviakrT",
}
`;

exports[`app/fetchAndTransformArticle 4835 2`] = `
"<section>
  <p>Forslag til årsplan 1P</p>
  <section class=\\"c-file-list\\">
    <h1 class=\\"c-file-list__heading\\">Filer</h1>
    <ul class=\\"c-file-list__files\\">
      <li class=\\"c-file-list__item\\">
        <a
          href=\\"https://test.api.ndla.no/files/103066/1p_arsplan_2013-2014_nynorsk.doc\\"
          class=\\"c-file-list__link\\"
          target=\\"_blank\\"
          aria-label=\\"Forslag til årsplan 1P (DOC)\\"
          aria-describedby=\\"testid_doc\\"
          ><svg
            fill=\\"currentColor\\"
            preserveAspectRatio=\\"xMidYMid meet\\"
            height=\\"1em\\"
            width=\\"1em\\"
            class=\\"c-icon\\"
            viewBox=\\"0 0 24 24\\"
            data-license=\\"Apache License 2.0\\"
            data-source=\\"Material Design\\"
            style=\\"vertical-align: middle\\"
          >
            <title>Download</title>
            <g>
              <path d=\\"M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z\\"></path>
              <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
            </g></svg
          ><span class=\\"c-file-list__link-text\\"
            ><span>Forslag til årsplan 1P (DOC)</span></span
          ><span
            class=\\"c-file-list__tooltip\\"
            aria-hidden=\\"true\\"
            role=\\"tooltip\\"
            id=\\"testid_doc\\"
            ><span class=\\"c-file-list__tooltip-text\\"
              >Last ned fil: 1p_arsplan_2013-2014_nynorsk.doc</span
            ></span
          ></a
        >
      </li>
      <li class=\\"c-file-list__item\\">
        <a
          href=\\"https://test.api.ndla.no/files/103066/1p_arsplan_2013-2014_nynorsk.odt\\"
          class=\\"c-file-list__link\\"
          target=\\"_blank\\"
          aria-label=\\"Forslag til årsplan 1P (ODT)\\"
          aria-describedby=\\"testid_odt\\"
          ><svg
            fill=\\"currentColor\\"
            preserveAspectRatio=\\"xMidYMid meet\\"
            height=\\"1em\\"
            width=\\"1em\\"
            class=\\"c-icon\\"
            viewBox=\\"0 0 24 24\\"
            data-license=\\"Apache License 2.0\\"
            data-source=\\"Material Design\\"
            style=\\"vertical-align: middle\\"
          >
            <title>Download</title>
            <g>
              <path d=\\"M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z\\"></path>
              <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
            </g></svg
          ><span class=\\"c-file-list__link-text\\"
            ><span>Forslag til årsplan 1P (ODT)</span></span
          ><span
            class=\\"c-file-list__tooltip\\"
            aria-hidden=\\"true\\"
            role=\\"tooltip\\"
            id=\\"testid_odt\\"
            ><span class=\\"c-file-list__tooltip-text\\"
              >Last ned fil: 1p_arsplan_2013-2014_nynorsk.odt</span
            ></span
          ></a
        >
      </li>
    </ul>
  </section>
  <figure data-sizetype=\\"full\\" class=\\"c-figure u-float-full\\" id=\\"0-testid\\">
    <h2>Forslag til årsplan 1P</h2>
    <iframe
      title=\\"Forslag til årsplan 1P\\"
      height=\\"1050\\"
      src=\\"https://test.api.ndla.no/files/103066/1p_arsplan_2013-2014_nynorsk.pdf\\"
    ></iframe>
  </figure>
</section>
"
`;
